File: output_alg/one_euro_z.cpp
#include "one_euro_z.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <stdexcept>

namespace output_alg {
namespace {

std::size_t DelayWindowSize(std::size_t delay_frames) {
    return delay_frames * 2 + 1;
}

std::size_t DelayedTargetIndex(std::size_t count, std::size_t delay_frames) {
    if (count == 0) {
        return 0;
    }
    if (count > delay_frames) {
        return count - 1 - delay_frames;
    }
    return 0;
}

}  // namespace

OneEuroZRealtimeFilter::OneEuroZRealtimeFilter(OneEuroZParameters params)
    : params_(params) {
    Validate(params_);
    Reset();
}

void OneEuroZRealtimeFilter::Reset() {
    initialized_ = false;
    last_raw_z_ = 0.0;
    filtered_z_ = 0.0;
    derivative_hat_ = 0.0;
    last_timestamp_.reset();
    raw_buffer_.clear();
    timestamp_buffer_.clear();
    history_.clear();
}

void OneEuroZRealtimeFilter::SetParameters(const OneEuroZParameters& params, bool reset) {
    Validate(params);
    params_ = params;
    if (reset) {
        Reset();
        return;
    }
    if (params_.history_size > 0) {
        while (history_.size() > params_.history_size) {
            history_.pop_front();
        }
    }
    const std::size_t max_delay_size = DelayWindowSize(params_.delay_frames);
    while (raw_buffer_.size() > max_delay_size) {
        raw_buffer_.pop_front();
        timestamp_buffer_.pop_front();
    }
}

Sn3DAlgorithm::RigidMatrix OneEuroZRealtimeFilter::Update(
    const Sn3DAlgorithm::RigidMatrix& rigid,
    OptionalDouble timestamp) {
    if (params_.delay_frames > 0) {
        if (timestamp.has_value() && last_timestamp_.has_value()) {
            const double dt = *timestamp - *last_timestamp_;
            if (dt <= 0.0 && params_.strict_timestamps) {
                throw std::invalid_argument("timestamps must be strictly increasing");
            }
        }
        initialized_ = true;
        PushDelayBuffer(rigid, timestamp);
        if (timestamp.has_value()) {
            last_timestamp_ = timestamp;
        }
        Sn3DAlgorithm::RigidMatrix filtered = DelayedWindowOutput();
        PushHistory(filtered);
        return filtered;
    }

    Sn3DAlgorithm::RigidMatrix filtered = rigid;
    const double raw_z = rigid.get_translation().z();

    if (!initialized_) {
        initialized_ = true;
        last_raw_z_ = raw_z;
        filtered_z_ = raw_z;
        derivative_hat_ = 0.0;
        last_timestamp_ = timestamp;
        filtered.get_translation().z() = raw_z;
        PushHistory(filtered);
        return filtered;
    }

    const double dt = DeltaTime(timestamp);
    const double derivative = (raw_z - last_raw_z_) / dt;
    const double derivative_alpha = LowpassAlpha(params_.d_cutoff, dt);
    derivative_hat_ =
        derivative_alpha * derivative + (1.0 - derivative_alpha) * derivative_hat_;

    const double effective_derivative =
        std::max(std::abs(derivative_hat_) - params_.derivative_deadband, 0.0);
    const double cutoff = params_.min_cutoff + params_.beta * effective_derivative;
    const double value_alpha = LowpassAlpha(cutoff, dt);
    filtered_z_ = value_alpha * raw_z + (1.0 - value_alpha) * filtered_z_;

    last_raw_z_ = raw_z;
    if (timestamp.has_value()) {
        last_timestamp_ = timestamp;
    }
    filtered.get_translation().z() = filtered_z_;
    PushHistory(filtered);
    return filtered;
}

TimedRigidResult OneEuroZRealtimeFilter::UpdateTimed(
    const Sn3DAlgorithm::RigidMatrix& rigid,
    OptionalDouble timestamp) {
    const auto started = std::chrono::steady_clock::now();
    TimedRigidResult result;
    result.rigid = Update(rigid, timestamp);
    result.elapsed_time_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::steady_clock::now() - started)
                                 .count();
    return result;
}

std::vector<Sn3DAlgorithm::RigidMatrix> OneEuroZRealtimeFilter::FilterTrajectory(
    const std::vector<Sn3DAlgorithm::RigidMatrix>& rigids,
    const std::vector<double>* timestamps,
    bool reset) {
    if (rigids.empty()) {
        throw std::invalid_argument("rigids must contain at least one frame");
    }
    if (timestamps != nullptr && timestamps->size() != rigids.size()) {
        throw std::invalid_argument("timestamps size must match rigids size");
    }
    if (reset) {
        Reset();
    }

    std::vector<Sn3DAlgorithm::RigidMatrix> output;
    output.reserve(rigids.size());
    for (std::size_t i = 0; i < rigids.size(); ++i) {
        if (timestamps != nullptr) {
            output.push_back(Update(rigids[i], (*timestamps)[i]));
        } else {
            output.push_back(Update(rigids[i]));
        }
    }
    return output;
}

TimedFilterTrajectoryResult OneEuroZRealtimeFilter::FilterTrajectoryTimed(
    const std::vector<Sn3DAlgorithm::RigidMatrix>& rigids,
    const std::vector<double>* timestamps,
    bool reset) {
    if (rigids.empty()) {
        throw std::invalid_argument("rigids must contain at least one frame");
    }
    if (timestamps != nullptr && timestamps->size() != rigids.size()) {
        throw std::invalid_argument("timestamps size must match rigids size");
    }
    if (reset) {
        Reset();
    }

    TimedFilterTrajectoryResult result;
    result.rigids.reserve(rigids.size());
    result.per_pose_time_ns.reserve(rigids.size());

    const auto started = std::chrono::steady_clock::now();
    for (std::size_t i = 0; i < rigids.size(); ++i) {
        const TimedRigidResult timed =
            timestamps != nullptr ? UpdateTimed(rigids[i], (*timestamps)[i]) : UpdateTimed(rigids[i]);
        result.rigids.push_back(timed.rigid);
        result.per_pose_time_ns.push_back(timed.elapsed_time_ns);
    }
    result.total_time_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::steady_clock::now() - started)
                               .count();
    FinalizePerPoseTimeNs(result.per_pose_time_ns, result.total_time_ns);
    return result;
}

Sn3DAlgorithm::RigidMatrix OneEuroZRealtimeFilter::FilterLatestFromHistory(
    const std::vector<Sn3DAlgorithm::RigidMatrix>& rigids,
    const std::vector<double>* timestamps) const {
    OneEuroZRealtimeFilter filter(params_);
    const auto filtered = filter.FilterTrajectory(rigids, timestamps, true);
    return filtered.back();
}

void OneEuroZRealtimeFilter::Validate(const OneEuroZParameters& params) {
    if (params.min_cutoff <= 0.0) {
        throw std::invalid_argument("min_cutoff must be positive");
    }
    if (params.beta < 0.0) {
        throw std::invalid_argument("beta must be >= 0");
    }
    if (params.d_cutoff <= 0.0) {
        throw std::invalid_argument("d_cutoff must be positive");
    }
    if (params.derivative_deadband < 0.0) {
        throw std::invalid_argument("derivative_deadband must be >= 0");
    }
    if (params.sample_rate_hz <= 0.0) {
        throw std::invalid_argument("sample_rate_hz must be positive");
    }
}

double OneEuroZRealtimeFilter::LowpassAlpha(double cutoff, double dt) {
    const double tau = 1.0 / (2.0 * std::acos(-1.0) * cutoff);
    return dt / (dt + tau);
}

double OneEuroZRealtimeFilter::DeltaTime(OptionalDouble timestamp) const {
    const double nominal = 1.0 / params_.sample_rate_hz;
    if (!timestamp.has_value() || !last_timestamp_.has_value()) {
        return nominal;
    }
    const double dt = *timestamp - *last_timestamp_;
    if (dt > 0.0) {
        return dt;
    }
    if (params_.strict_timestamps) {
        throw std::invalid_argument("timestamps must be strictly increasing");
    }
    return nominal;
}

void OneEuroZRealtimeFilter::PushDelayBuffer(
    const Sn3DAlgorithm::RigidMatrix& rigid,
    OptionalDouble timestamp) {
    raw_buffer_.push_back(rigid);
    timestamp_buffer_.push_back(timestamp);
    const std::size_t max_size = DelayWindowSize(params_.delay_frames);
    while (raw_buffer_.size() > max_size) {
        raw_buffer_.pop_front();
        timestamp_buffer_.pop_front();
    }
}

Sn3DAlgorithm::RigidMatrix OneEuroZRealtimeFilter::DelayedWindowOutput() const {
    if (raw_buffer_.empty()) {
        throw std::invalid_argument("delay buffer is empty");
    }
    const std::size_t target_index =
        DelayedTargetIndex(raw_buffer_.size(), params_.delay_frames);
    Sn3DAlgorithm::RigidMatrix filtered = raw_buffer_[target_index];
    filtered.get_translation().z() = SmoothedDelayedZ();
    return filtered;
}

double OneEuroZRealtimeFilter::SmoothedDelayedZ() const {
    const std::size_t count = raw_buffer_.size();
    const std::size_t target_index = DelayedTargetIndex(count, params_.delay_frames);
    if (count <= 1) {
        return raw_buffer_[target_index].get_translation().z();
    }

    const double dt = BufferMeanDt();
    const double derivative_hat = EstimateWindowDerivative(target_index);
    const double effective_derivative =
        std::max(std::abs(derivative_hat) - params_.derivative_deadband, 0.0);
    const double cutoff = params_.min_cutoff + params_.beta * effective_derivative;
    const double value_alpha = LowpassAlpha(cutoff, dt);
    const double decay = 1.0 - value_alpha;

    double weighted_sum = 0.0;
    double weight_sum = 0.0;
    for (std::size_t index = 0; index < count; ++index) {
        const double distance =
            std::abs(static_cast<double>(index) - static_cast<double>(target_index));
        const double weight = std::pow(decay, distance);
        weighted_sum += weight * raw_buffer_[index].get_translation().z();
        weight_sum += weight;
    }
    if (weight_sum <= 0.0) {
        return raw_buffer_[target_index].get_translation().z();
    }
    return weighted_sum / weight_sum;
}

double OneEuroZRealtimeFilter::BufferDt(std::size_t previous_index, std::size_t index) const {
    const double nominal = 1.0 / params_.sample_rate_hz;
    if (previous_index >= timestamp_buffer_.size() || index >= timestamp_buffer_.size()) {
        return nominal;
    }
    const OptionalDouble& previous = timestamp_buffer_[previous_index];
    const OptionalDouble& current = timestamp_buffer_[index];
    if (!previous.has_value() || !current.has_value()) {
        return nominal;
    }
    const double dt = *current - *previous;
    if (dt > 0.0) {
        return dt;
    }
    if (params_.strict_timestamps) {
        throw std::invalid_argument("timestamps must be strictly increasing");
    }
    return nominal;
}

double OneEuroZRealtimeFilter::BufferMeanDt() const {
    if (raw_buffer_.size() <= 1) {
        return 1.0 / params_.sample_rate_hz;
    }
    double sum = 0.0;
    std::size_t count = 0;
    for (std::size_t index = 1; index < raw_buffer_.size(); ++index) {
        const double dt = BufferDt(index - 1, index);
        if (dt > 0.0) {
            sum += dt;
            ++count;
        }
    }
    if (count == 0) {
        return 1.0 / params_.sample_rate_hz;
    }
    return sum / static_cast<double>(count);
}

double OneEuroZRealtimeFilter::EstimateWindowDerivative(std::size_t target_index) const {
    if (raw_buffer_.size() <= 1) {
        return 0.0;
    }

    const double derivative_alpha = LowpassAlpha(params_.d_cutoff, BufferMeanDt());
    const double derivative_decay = 1.0 - derivative_alpha;
    double weighted_sum = 0.0;
    double weight_sum = 0.0;
    for (std::size_t index = 1; index < raw_buffer_.size(); ++index) {
        const double dt = BufferDt(index - 1, index);
        const double previous_z = raw_buffer_[index - 1].get_translation().z();
        const double current_z = raw_buffer_[index].get_translation().z();
        const double derivative = (current_z - previous_z) / dt;
        const double step_center = static_cast<double>(index) - 0.5;
        const double distance = std::abs(step_center - static_cast<double>(target_index));
        const double weight = std::pow(derivative_decay, distance);
        weighted_sum += weight * derivative;
        weight_sum += weight;
    }
    if (weight_sum <= 0.0) {
        return 0.0;
    }
    return weighted_sum / weight_sum;
}

void OneEuroZRealtimeFilter::PushHistory(const Sn3DAlgorithm::RigidMatrix& rigid) {
    history_.push_back(rigid);
    if (params_.history_size > 0) {
        while (history_.size() > params_.history_size) {
            history_.pop_front();
        }
    }
}

std::vector<Sn3DAlgorithm::RigidMatrix> FilterOneEuroZTrajectory(
    const std::vector<Sn3DAlgorithm::RigidMatrix>& rigids,
    const std::vector<double>* timestamps,
    OneEuroZParameters params) {
    OneEuroZRealtimeFilter filter(params);
    return filter.FilterTrajectory(rigids, timestamps, true);
}

TimedFilterTrajectoryResult FilterOneEuroZTrajectoryTimed(
    const std::vector<Sn3DAlgorithm::RigidMatrix>& rigids,
    const std::vector<double>* timestamps,
    OneEuroZParameters params) {
    OneEuroZRealtimeFilter filter(params);
    return filter.FilterTrajectoryTimed(rigids, timestamps, true);
}

Sn3DAlgorithm::RigidMatrix FilterOneEuroZLatestFromHistory(
    const std::vector<Sn3DAlgorithm::RigidMatrix>& rigids,
    const std::vector<double>* timestamps,
    OneEuroZParameters params) {
    OneEuroZRealtimeFilter filter(params);
    return filter.FilterLatestFromHistory(rigids, timestamps);
}

}  // namespace output_alg

File: output_alg/one_euro_z.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Iterable

import numpy as np
from numpy.typing import ArrayLike, NDArray


PoseArray = NDArray[np.float64]


@dataclass
class OneEuroZParameters:
    """Runtime parameters for Z-only One Euro filtering.

    Lower ``min_cutoff`` gives stronger static denoising and more lag.
    Higher ``beta`` raises the cutoff during motion, reducing lag at the cost
    of letting more motion-time noise through.
    ``derivative_deadband`` ignores small filtered Z velocity before adapting
    the cutoff, which keeps static depth noise from weakening the denoising.
    """

    min_cutoff: float = 1.0
    beta: float = 10.0
    d_cutoff: float = 8.0
    derivative_deadband: float = 0.02
    sample_rate_hz: float = 100.0
    history_size: int = 0
    delay_frames: int = 0
    strict_timestamps: bool = False

    def validate(self) -> None:
        if self.min_cutoff <= 0:
            raise ValueError("min_cutoff must be positive")
        if self.beta < 0:
            raise ValueError("beta must be >= 0")
        if self.d_cutoff <= 0:
            raise ValueError("d_cutoff must be positive")
        if self.derivative_deadband < 0:
            raise ValueError("derivative_deadband must be >= 0")
        if self.sample_rate_hz <= 0:
            raise ValueError("sample_rate_hz must be positive")
        if self.history_size < 0:
            raise ValueError("history_size must be >= 0")
        if self.delay_frames < 0:
            raise ValueError("delay_frames must be >= 0")


class OneEuroZRealtimeFilter:
    """Realtime Z-only One Euro filter for 4x4 trajectory poses.

    Input and output poses have shape ``(4, 4)``. Only ``pose[2, 3]`` is
    filtered; X/Y translation and all rotation matrix entries are preserved.
    """

    def __init__(self, params: OneEuroZParameters | None = None) -> None:
        self.params = OneEuroZParameters() if params is None else params
        self.params.validate()
        self.reset()

    def reset(self) -> None:
        self._initialized = False
        self._last_raw_z = 0.0
        self._filtered_z = 0.0
        self._derivative_hat = 0.0
        self._last_timestamp: float | None = None
        self._history: deque[PoseArray] = deque(maxlen=self.params.history_size or None)
        self._raw_buffer: deque[PoseArray] = deque(maxlen=_delay_window_size(self.params.delay_frames))
        self._timestamp_buffer: deque[float | None] = deque(
            maxlen=_delay_window_size(self.params.delay_frames)
        )

    def set_parameters(self, params: OneEuroZParameters, *, reset: bool = False) -> None:
        params.validate()
        self.params = params
        if reset:
            self.reset()
        else:
            self._history = deque(self._history, maxlen=self.params.history_size or None)
            self._raw_buffer = deque(
                self._raw_buffer,
                maxlen=_delay_window_size(self.params.delay_frames),
            )
            self._timestamp_buffer = deque(
                self._timestamp_buffer,
                maxlen=_delay_window_size(self.params.delay_frames),
            )

    @property
    def history(self) -> list[PoseArray]:
        return [pose.copy() for pose in self._history]

    def update(self, pose: ArrayLike, timestamp: float | None = None) -> PoseArray:
        """Filter one incoming pose and return one display/output pose."""

        pose_arr = _as_pose(pose)
        filtered = pose_arr.copy()
        raw_z = float(pose_arr[2, 3])

        if self.params.delay_frames > 0:
            self._push_delay_frame(pose_arr, timestamp)
            target_index = _delayed_target_index(len(self._raw_buffer), self.params.delay_frames)
            filtered = self._raw_buffer[target_index].copy()
            filtered[2, 3] = self._smoothed_delayed_z(target_index)
            if timestamp is not None:
                self._last_timestamp = float(timestamp)
            self._initialized = True
            self._history.append(filtered.copy())
            return filtered

        if not self._initialized:
            self._initialized = True
            self._last_raw_z = raw_z
            self._filtered_z = raw_z
            self._derivative_hat = 0.0
            self._last_timestamp = None if timestamp is None else float(timestamp)
            filtered[2, 3] = raw_z
            self._history.append(filtered.copy())
            return filtered

        dt = self._dt(timestamp)
        derivative = (raw_z - self._last_raw_z) / dt
        derivative_alpha = _lowpass_alpha(self.params.d_cutoff, dt)
        self._derivative_hat = (
            derivative_alpha * derivative + (1.0 - derivative_alpha) * self._derivative_hat
        )

        effective_derivative = max(abs(self._derivative_hat) - self.params.derivative_deadband, 0.0)
        cutoff = self.params.min_cutoff + self.params.beta * effective_derivative
        value_alpha = _lowpass_alpha(cutoff, dt)
        self._filtered_z = value_alpha * raw_z + (1.0 - value_alpha) * self._filtered_z
        self._last_raw_z = raw_z
        if timestamp is not None:
            self._last_timestamp = float(timestamp)

        filtered[2, 3] = self._filtered_z
        self._history.append(filtered.copy())
        return filtered

    def _push_delay_frame(self, pose: PoseArray, timestamp: float | None) -> None:
        if timestamp is not None and self._last_timestamp is not None:
            dt = float(timestamp) - self._last_timestamp
            if dt <= 0 and self.params.strict_timestamps:
                raise ValueError("timestamps must be strictly increasing")
        self._raw_buffer.append(pose.copy())
        self._timestamp_buffer.append(None if timestamp is None else float(timestamp))

    def _smoothed_delayed_z(self, target_index: int) -> float:
        if len(self._raw_buffer) <= 1:
            return float(self._raw_buffer[target_index][2, 3])

        dt = self._buffer_mean_dt()
        derivative_hat = self._estimate_window_derivative(target_index)
        effective_derivative = max(abs(derivative_hat) - self.params.derivative_deadband, 0.0)
        cutoff = self.params.min_cutoff + self.params.beta * effective_derivative
        value_alpha = _lowpass_alpha(cutoff, dt)
        decay = 1.0 - value_alpha

        weighted_sum = 0.0
        weight_sum = 0.0
        for index, pose in enumerate(self._raw_buffer):
            weight = decay ** abs(index - target_index)
            weighted_sum += weight * float(pose[2, 3])
            weight_sum += weight
        if weight_sum <= 0.0:
            return float(self._raw_buffer[target_index][2, 3])
        return float(weighted_sum / weight_sum)

    def _buffer_dt(self, previous_index: int, index: int) -> float:
        nominal = 1.0 / self.params.sample_rate_hz
        try:
            previous = self._timestamp_buffer[previous_index]
            current = self._timestamp_buffer[index]
        except IndexError:
            return nominal
        if previous is None or current is None:
            return nominal
        dt = float(current) - float(previous)
        if dt > 0:
            return dt
        if self.params.strict_timestamps:
            raise ValueError("timestamps must be strictly increasing")
        return nominal

    def _buffer_mean_dt(self) -> float:
        if len(self._raw_buffer) <= 1:
            return 1.0 / self.params.sample_rate_hz
        values = [self._buffer_dt(index - 1, index) for index in range(1, len(self._raw_buffer))]
        valid = [value for value in values if value > 0.0]
        if not valid:
            return 1.0 / self.params.sample_rate_hz
        return float(np.mean(valid))

    def _estimate_window_derivative(self, target_index: int) -> float:
        if len(self._raw_buffer) <= 1:
            return 0.0
        derivative_alpha = _lowpass_alpha(self.params.d_cutoff, self._buffer_mean_dt())
        derivative_decay = 1.0 - derivative_alpha
        weighted_sum = 0.0
        weight_sum = 0.0
        for index in range(1, len(self._raw_buffer)):
            dt = self._buffer_dt(index - 1, index)
            previous_z = float(self._raw_buffer[index - 1][2, 3])
            current_z = float(self._raw_buffer[index][2, 3])
            derivative = (current_z - previous_z) / dt
            step_center = float(index) - 0.5
            weight = derivative_decay ** abs(step_center - float(target_index))
            weighted_sum += weight * derivative
            weight_sum += weight
        if weight_sum <= 0.0:
            return 0.0
        return float(weighted_sum / weight_sum)

    def filter_trajectory(
        self,
        poses: ArrayLike,
        timestamps: Iterable[float] | None = None,
        *,
        reset: bool = True,
    ) -> PoseArray:
        """Filter every input frame and return a filtered ``(N, 4, 4)`` array."""

        pose_arr = _as_pose_sequence(poses)
        time_values = _timestamps(timestamps, pose_arr.shape[0])
        if reset:
            self.reset()
        output = np.empty_like(pose_arr)
        for idx, pose in enumerate(pose_arr):
            ts = None if time_values is None else float(time_values[idx])
            output[idx] = self.update(pose, ts)
        return output

    def filter_latest_from_history(
        self,
        poses: ArrayLike,
        timestamps: Iterable[float] | None = None,
    ) -> PoseArray:
        """Filter a bounded history and return only the latest filtered pose."""

        filtered = self.filter_trajectory(poses, timestamps, reset=True)
        if filtered.shape[0] == 0:
            raise ValueError("poses must contain at least one frame")
        return filtered[-1]

    def _dt(self, timestamp: float | None) -> float:
        nominal = 1.0 / self.params.sample_rate_hz
        if timestamp is None or self._last_timestamp is None:
            return nominal
        dt = float(timestamp) - self._last_timestamp
        if dt > 0:
            return dt
        if self.params.strict_timestamps:
            raise ValueError("timestamps must be strictly increasing")
        return nominal


def filter_trajectory(
    poses: ArrayLike,
    timestamps: Iterable[float] | None = None,
    params: OneEuroZParameters | None = None,
) -> PoseArray:
    """Stateless helper for batch-style per-frame output."""

    return OneEuroZRealtimeFilter(params).filter_trajectory(poses, timestamps)


def filter_latest_from_history(
    poses: ArrayLike,
    timestamps: Iterable[float] | None = None,
    params: OneEuroZParameters | None = None,
) -> PoseArray:
    """Stateless helper for UI display from a recent pose history."""

    return OneEuroZRealtimeFilter(params).filter_latest_from_history(poses, timestamps)


def _as_pose(pose: ArrayLike) -> PoseArray:
    pose_arr = np.asarray(pose, dtype=float)
    if pose_arr.shape != (4, 4):
        raise ValueError(f"pose must have shape (4, 4), got {pose_arr.shape}")
    return pose_arr


def _as_pose_sequence(poses: ArrayLike) -> PoseArray:
    pose_arr = np.asarray(poses, dtype=float)
    if pose_arr.ndim != 3 or pose_arr.shape[1:] != (4, 4):
        raise ValueError(f"poses must have shape (N, 4, 4), got {pose_arr.shape}")
    if pose_arr.shape[0] == 0:
        raise ValueError("poses must contain at least one frame")
    return pose_arr


def _timestamps(timestamps: Iterable[float] | None, count: int) -> PoseArray | None:
    if timestamps is None:
        return None
    values = np.asarray(list(timestamps), dtype=float)
    if values.shape != (count,):
        raise ValueError(f"timestamps must have shape ({count},), got {values.shape}")
    return values


def _lowpass_alpha(cutoff: float, dt: float) -> float:
    tau = 1.0 / (2.0 * np.pi * cutoff)
    return float(dt / (dt + tau))


def _delay_window_size(delay_frames: int) -> int | None:
    if delay_frames <= 0:
        return None
    return 2 * int(delay_frames) + 1


def _delayed_target_index(count: int, delay_frames: int) -> int:
    if count <= 0:
        return 0
    if count > delay_frames:
        return count - 1 - delay_frames
    return 0
